staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/convert_test.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	fuzz "github.com/google/gofuzz"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/json"
+)
+
+func TestStructuralRoundtripOrError(t *testing.T) {
+	f := fuzz.New()
+	seed := time.Now().UnixNano()
+	t.Logf("seed = %v", seed)
+	//seed = int64(1549012506261785182)
+	f.RandSource(rand.New(rand.NewSource(seed)))
+	f.Funcs(
+		func(s *apiextensions.JSON, c fuzz.Continue) {
+			*s = apiextensions.JSON(map[string]interface{}{"foo": float64(42.2)})
+		},
+		func(s *apiextensions.JSONSchemaPropsOrArray, c fuzz.Continue) {
+			c.FuzzNoCustom(s)
+			if s.Schema != nil {
+				s.JSONSchemas = nil
+			} else if s.JSONSchemas == nil {
+				s.Schema = &apiextensions.JSONSchemaProps{}
+			}
+		},
+		func(s *apiextensions.JSONSchemaPropsOrBool, c fuzz.Continue) {
+			c.FuzzNoCustom(s)
+			if s.Schema != nil {
+				s.Allows = false
+			}
+		},
+		func(s **string, c fuzz.Continue) {
+			c.FuzzNoCustom(s)
+			if *s != nil && **s == "" {
+				*s = nil
+			}
+		},
+	)
+
+	f.MaxDepth(2)
+	f.NilChance(0.5)
+
+	for i := 0; i < 10000; i++ {
+		// fuzz a random field in JSONSchemaProps
+		origSchema := &apiextensions.JSONSchemaProps{}
+		x := reflect.ValueOf(origSchema).Elem()
+		n := rand.Intn(x.NumField())
+		if name := x.Type().Field(n).Name; name == "Example" || name == "ExternalDocs" || name == "Nullable" {
+			// we drop these intentionally, or in case of Nullable know that it does not roundtrip.
```

Comment:
comment (and ideally link to issue) why nullable doesn't roundtrip... I can't remember why it doesn't

Reply:
I added the `nullable:true` case with non-empty type to the test. And I explained the remaining empty type case.